mri/reconstructors/calibrationless.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+"""
+This implements calibrationless reconstruction with different proximities
+"""
+# System import
+import warnings
+
+from .base import ReconstructorBase
+from ..operators import GradAnalysis, GradSynthesis, WaveletN
+
+# Third party import
+from modopt.opt.proximity import SparseThreshold
+from modopt.opt.linear import Identity
+
+
+class CalibrationlessReconstructor(ReconstructorBase):
+    """ This class implements a regularized calibrationless reconstruction.
+    For the Analysis case, finds the solution for x of:
+        (1/2) * sum(||F x_l - y_l||^2_2, n_coils) +
+                    mu * H(W x_l)
+    For the Synthesis case, finds the solution of:
+        (1/2) * sum(||F Wt alpha_l - y_l||^2_2, n_coils) +
+                    mu * H(alpha_l)
+    Parameters
+    ----------
+    fourier_op: object of class FFT, NonCartesianFFT or Stacked3DNFFT in
+                mri.operators
+        Defines the fourier operator F in the above equation.
+    linear_op: object, (optional, default None)
+        Defines the linear sparsifying operator W. This must operate on x and
+        have 2 functions, op(x) and adj_op(coeff) which implements the
+        operator and adjoint operator. For wavelets, this can be object of
+        class WaveletN or WaveletUD2 from mri.operators .
+        If None, sym8 wavelet with nb_scale=3 is chosen.
+    regularizer_op: operator, (optional default None)
+        Defines the regularization operator for the regularization function H.
+        If None, the  regularization chosen is Identity and the optimization
+        turns to gradient descent.
+    gradient_formulation: str between 'analysis' or 'synthesis',
+        default 'synthesis'
+        defines the formulation of the image model which defines the gradient.
+    n_jobs : int, default 1
+        The number of cores to be used for faster reconstruction
+    verbose: int, default 0
+        Verbosity level.
+            1 => Print basic debug information
+            5 => Print all initialization information
+            20 => Calculate cost at the end of each iteration.
+                NOTE : This is computationally intensive.
+            30 => Print the debug information of operators if defined by class
+    **kwargs : Extra keyword arguments
+        for gradient initialization:
+            Please refer to mri.operators.gradient.base for information
```

Comment:
they are not only for the gradient right (since they also include the regularizer). Maybe it would be nice to point which specific gradient don't you think? I guess they have different kwargs.

Reply:
Well its just Lipchitz constant args, and thus I pointed to base class to keep it uniform,
I was wondering how to add the regularizer_op doc here. 
Would having regularizer_op definition make sense here?

Reply:
Ok makes sense.
Yes it's not only the gradient args, we also have regularizer op, so I would have it here.

Reply:
Cool..